File: python/algo/202603/algo3762.py
from typing import List, Tuple
from __future__ import annotations

class Node:
    __slots__ = 'l', 'r', 'lo', 'ro', 'cnt', 'sum'

    def __init__(self, l:int, r: int, lo:Node=None, ro:Node = None, cnt:int=0, sum:int =0):
        self.l = l
        self.r = r
        self.lo = lo
        self.ro = ro
        self.cnt = cnt
        self.sum = sum

    def maintain(self):
        self.cnt = self.lo.cnt + self.ro.cnt
        self.sum = self.lo.sum + self.ro.sum

    @staticmethod
    def build(l: int, r: int) -> Node:
        o = Node(l, r)
        if l == r: return o
        mid = (l + r) // 2
        o.lo = Node.build(l, mid)
        o.ro = Node.build(mid+1, r)

        return o

    # i 添加 al
    def add(self, rank:int, val:int) -> Node:
        o = Node(self.l, self.r, self.lo, self.ro, self.cnt, self.sum)
        if o.l == o.r:
            o.cnt +=1
            o.sum += val
            return o
        mid = (o.l + o.r) // 2
        if rank <= mid: o.lo = o.lo.add(rank, val)
        else: o.ro = o.ro.add(rank, val)
        o.maintain()
        return o

    def kth(self, prev: Node, k: int) -> int:
        if self.l == self.r: return self.l

        cnt_l = self.lo.cnt - prev.lo.cnt
        if k <= cnt_l: return self.lo.kth(prev.lo, k)
        else: return self.ro.kth(prev.ro, k-cnt_l)
    
    def query(self, prev: Node, rank: int) -> Tuple[int, int]:
        if self.r <= rank: return self.cnt - prev.cnt, self.sum - prev.sum
        cnt, sum_ = self.lo.query(prev.lo, rank)
        mid = (self.l + self.r) // 2
        if rank > mid:
            cnt_r, sum_r = self.ro.query(prev.ro, rank)
            cnt += cnt_r
            sum_ += sum_r
        return cnt, sum_

class Solution:
    def minOperations(self, nums: List[int], k: int, queries: List[List[int]]) -> List[int]:
        n = len(nums)
        left = [0] * n

        for i in range(1, n):
            left[i] = left[i-1] if nums[i] % k == nums[i-1] % k else i
        
        uni_nums = sorted(set(nums))
        val_to_rank = {num: i for i, num in enumerate(uni_nums)}

        t = [None] * (n + 1)
        t[0] = Node.build(0, len(uni_nums) - 1)

        for i, val in enumerate(nums):
            rank = val_to_rank[val]
            t[i+1] = t[i].add(rank, val)

        ans = []
        for l, r in queries:
            if left[r] > l: 
                ans.append(-1)
                continue
            
            r += 1
            seg_len = r - l
            k_rank = t[r].kth(t[l], seg_len // 2 + 1)
            median = uni_nums[k_rank]
            
            total = t[r].sum - t[l].sum
            cnt_l, sum_l = t[r].query(t[l], k_rank)
            
            s = median * cnt_l - sum_l
            s += total - sum_l - median * (seg_len - cnt_l)
            ans.append(s // k)
        return ans
        

File: cplus/leetcode/202603/algo3762.cpp
#include <bits/stdc++.h>
using namespace std;

// 主席树，持久化线段树 模板 独立类
// Dynamic Segment Tree  (Dynamic Node Allocation Segment Tree)
class Node {
    int l, r;            //  区间
    Node *left, *right;  //  左右子树
    int cnt;
    void maintain() {
        cnt = left->cnt + right->cnt;
        sum = left->sum + right->sum;
    }
public:
    long long sum;
    Node(int l, int r, Node* left = nullptr, Node* right = nullptr, int cnt = 0, long long sum = 0)
        : l(l), r(r), left(left), right(right), cnt(cnt), sum(sum) {}

    static Node* build(int l, int r) {
        Node* node = new Node(l, r);
        if (l != r) {
            int m = (l + r) / 2;
            node->left = build(l, m);
            node->right = build(m+1, r);
        }
        return node;
    }

    Node* add(int i, int val) {
        Node * cur = new Node(l, r, left, right, cnt, sum);
        if (l == r) {
            cur->cnt += 1; cur->sum += val;
            return cur;
        }
        int m = (l + r) / 2; 
        if (i <= m) cur->left = cur->left->add(i, val);
        else cur->right = cur->right->add(i, val);
        cur->maintain();
        return cur;
    }

    int kth(Node* prev, int k) {
        if (l == r) return l;
        // 从根节点的视角，包括所有区间元素，因此应大于 k，因此先比较左子树的区间元素数量
        int rem = left->cnt - prev->left->cnt;
        if (rem >= k) return left->kth(prev->left, k);
        else return right->kth(prev->right, k - rem);
    }

    pair<int, long long> query(Node* prev, int m_rank) {
        if (r <= m_rank) return {cnt - prev->cnt, sum - prev->sum};
        auto [cnt, sum] = left->query(prev->left, m_rank);
        int m = (l + r) / 2;
        if (m_rank > m) {
            auto [r_cnt, r_sum] = right->query(prev->right, m_rank);
            cnt += r_cnt; sum += r_sum;
        }
        return {cnt, sum};
    }

};

// 四、数学贪心 - 4.5 中位数贪心
class Solution {
public:
    vector<long long> minOperations(vector<int>& nums, int k, vector<vector<int>>& queries) {
        // 复习
        int n = nums.size();
        vector<int> left(n);
        for (int i=1; i<n; i++) left[i] = nums[i] % k == nums[i-1] % k? left[i-1] : i;

        // 离散化 nums
        vector<int> uni_nums = nums;
        sort(uni_nums.begin(), uni_nums.end());
        uni_nums.erase(ranges::unique(uni_nums).begin(), uni_nums.end());   // distinct

        int m = uni_nums.size();
        vector<Node*> pst(n+1);  // persistent segment tree
        pst[0] = Node::build(0, m-1);   // 线段树的值域 [0, m)
        for (int i=0; i<n; i++) {
            int j = ranges::lower_bound(uni_nums, nums[i]) - uni_nums.begin();
            pst[i+1] = pst[i]->add(j, nums[i]);
        }

        vector<long long> ans;
        for (auto & q: queries) {
            auto l = q[0], r = q[1] + 1;
            if (left[r-1] > l) {ans.emplace_back(-1); continue;}

            int len = r - l;      // number of elements in interval
            int m_rank = pst[r]->kth(pst[l], len / 2 + 1);  // 中位数
            long long median = uni_nums[m_rank];
            // 求所有数与中位数的距离和，利用前缀和
            long long total = pst[r]->sum - pst[l]->sum;
            auto [l_cnt, l_sum] = pst[r]->query(pst[l], m_rank);
            
            long long res = median * l_cnt - l_sum;
            res += total - l_sum - median * (len-l_cnt);
            ans.emplace_back(res / k);
        } 
        return ans;
    }   
};